File: agent-world-lab/backend/src/AgentWorldLab.Infrastructure/Services/ToolAuthorizationService.cs
using AgentWorldLab.Application.Abstractions;
using AgentWorldLab.Application.Models;
using AgentWorldLab.Domain.Models;

namespace AgentWorldLab.Infrastructure.Services;

public sealed class ToolAuthorizationService : IToolAuthorizationService
{
    public bool IsToolAllowed(CurrentUser user, ToolDefinition tool, string environment)
    {
        _ = environment;

        if (string.IsNullOrWhiteSpace(user.Email))
        {
            return false;
        }

        return user.Roles.Any(role => tool.AllowedRoles.Contains(role, StringComparer.OrdinalIgnoreCase));
    }
}

File: agent-world-lab/backend/python/app/infrastructure/services/tool_authorization_service.py
from __future__ import annotations

from app.application.interfaces import IToolAuthorizationService
from app.application.models import CurrentUser
from app.domain.models import ToolDefinition


class ToolAuthorizationService(IToolAuthorizationService):
    def is_tool_allowed(self, user: CurrentUser, tool: ToolDefinition, environment: str | None) -> bool:
        allowed = {role.lower() for role in tool.allowed_roles}
        return any(role.lower() in allowed for role in user.roles)
